### DISCUSSION TITLE
How to set cookies during server-side rendering in App router?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/next.js#72370] [Help] [answered]
### Summary

The only way I can foresee setting cookies is by waiting for the frontend to load, then making an HTTP API request, setting the cookies in the `route.ts` handler, and then boom the frontend has a server-side cookie set. Now every subsequent frontend request that is made will send the cookies.

How can we skip the step of having to call the API endpoint from the frontend at the get-go, and instead make it so you can set the cookies during the view rendering server-side?

Something like:

```ts
'use server'

export default function Page() {
  // Set the cookie with required attributes for cross-origin sharing
  cookies().set({
    name: "thename",
    value: "thevalue",
    path: "/",
    domain: "thedomain.com", // Only necessary if you want the cookie for thedomain.com
    httpOnly: true,
    secure: true,
    sameSite: "none",
  });

  return <div>hello world</div>
}
```

Is it possible?

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6030
Binaries:
  Node: 20.10.0
  npm: 10.4.0
  Yarn: 1.22.19
  pnpm: 9.11.0
Relevant Packages:
  next: 14.0.4
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 4.9.5
Next.js Config:
  output: N/A
```

**Comment by icyJoseph** [comment, marked as answer]
No, here you are mixing up a few things.

Adding `use server` is a way for the client to call a piece of code in the server. 

It doesn't make your Page a server action.

The problem is that streaming requires the headers of the response to be sent to the client upfront. This includes the `Set-Cookie` header of course. By the time you are rendering your component, you are already at the streaming the response body stage, and you can't include any headers there.

One way to do this is middleware, or as you say, do some frontend acrobatics to get that cookie set.

**Reply by margaridafp** [reply]
My backend API is set up to send a new authentication cookie, using a sliding expiration model to keep the user authenticated as long as they remain active on the application.
I'm currently fetching data from this backend API in both server and client-side components. However, I'm unsure of the best way to handle updating the authentication cookie on the server-side components.

Do you have any guidance or recommendations on the best way to manage this authentication cookie update process? Nextjs version 14.2.13 and app router.
